io/include/pcl/io/real_sense_2_grabber.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2018-, Open Perception, Inc.
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_IO_REAL_SENSE_2_GRABBER_H
+#define PCL_IO_REAL_SENSE_2_GRABBER_H
+
+#include <pcl/io/boost.h>
+#include <pcl/io/grabber.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+
+#include <librealsense2/rs.hpp>
+
+namespace pcl
+{
+
+  /** \brief Grabber for Intel Realsense 2 SDK devices (D400 series)
+  * \author Patrick Abadi <patrickabadi@gmail.com>, Daniel Packard <pack3754@gmail.com>
+  * \ingroup io
+  */
+  class PCL_EXPORTS RealSense2Grabber : public pcl::Grabber
+  {
+  public:
+    /** \brief Constructor
+    * \param[in] serial_number optional to start up a specific device
+    */
+    RealSense2Grabber (const uint32_t serial_number = 0);
```

Comment:
The serial number requires 64-bits width (<code>uint64_t</code>).

Reply:
I am not sure if a `uint` is an appropriate representation. Even if serial number is digits-only (did anyone guarantee that?), it can be the case that the first digit is zero. (I actually have an older generation Intel camera that has serial number "047140085403".) The SDK uses strings, therefore we should use strings.

Reply:
> it can be the case that the first digit is zero. (I actually have an older generation Intel camera that has serial number "047140085403".)

Oops! I didn't consider that case.
That case should be considered.

Reply:
Receive serial number and bag file path as single arguments?
```cpp
RealSense2Grabber (const std::string& serial_number_or_bag_file = "");
```
```cpp
RealSense2Grabber::start()
{
  running = true;

  rs2::config cfg;

  // If string contain ".bag", capture from file
  if (serial_number_or_bag_file.rfind (".bag") == 4)
  {
    cfg.enable_device_from_file (serial_number_or_bag_file);
  }
  // otherwise, capture from sensor
  else
  {
    if (serial_number_or_bag_file)
      cfg.enable_device (serial_number_or_bag_file);

    cfg.enable_stream (RS2_STREAM_COLOR, deviceWidth, deviceHeight, RS2_FORMAT_RGB8, targetFps);
    cfg.enable_stream (RS2_STREAM_DEPTH, deviceWidth, deviceHeight, RS2_FORMAT_ANY, targetFps);
    cfg.enable_stream (RS2_STREAM_INFRARED, deviceWidth, deviceHeight, RS2_FORMAT_ANY, targetFps);
  }

  pipe.start(cfg);

  thread = std::thread(&RealSense2Grabber::threadFunction, this);
}
```

Reply:
@daniel-packard and I had a discussion on this and felt that using static factory method constructors is a nicer way to expose the serial number and file names.  If you disagree please let us know

Reply:
I think a constructor with a single string argument is very straightforward and easy to use API (given a good documentation comment is supplied).

Think about it this way. Typically, it is the application user who decides which camera he wants to use (options being "any camera", "specific camera", "bag file"). Thus, it is convenient to have e.g. a command-line argument "data source", where the user will either pass an empty string, or serial number, or a path to a bag file. If we adopt your factory method, then the application developer will need to foresee all these possibilities, detect them, and call appropriate factory methods.

Reply:
@taketwo Sounds good and I've updated the code with the single constructor.